python: custom headers not working

Hi, I was trying to play around with the Python ADBC driver for FlightSQL to see if I could connect to a service we're building. Our service requires some custom headers, but when I try to set them it breaks.

Example code:

```
from adbc_driver_flightsql.dbapi import connect


def main():
    conn = connect(
        "grpc://localhost:9876",
        conn_kwargs={
            "adbc.flight.sql.rpc.call_header.customheader": "1",
        },
    )
    cur = conn.cursor()
    cur.execute("select * from table")
    print(cur.fetch_arrow_table())


main()
```

error:
```
panic: interface conversion: interface is nil, not adbc.PostInitOptions

goroutine 17 [running, locked to thread]:
main.FlightSQLConnectionSetOption(0x1004537f8, 0x0?, 0x100453701?, 0x14000002680?)
        /Users/runner/work/arrow-adbc/arrow-adbc/adbc/go/adbc/pkg/flightsql/driver.go:249 +0x70
```

I don't have much experience with go, much less with Python bindings on top of go, so I don't know how to even start debugging this. Any help would be greatly appreciated!

--- a/go/adbc/pkg/_tmpl/driver.go.tmpl
+++ b/go/adbc/pkg/_tmpl/driver.go.tmpl
@@ -243,7 +243,8 @@ func {{.Prefix}}DatabaseRelease(db *C.struct_AdbcDatabase, err *C.struct_AdbcErr
 }
 
 type cConn struct {
-	cnxn adbc.Connection
+	cnxn     adbc.Connection
+	initArgs map[string]string
 }
 
 func checkConnAlloc(cnxn *C.struct_AdbcConnection, err *C.struct_AdbcError, fname string) bool {
@@ -308,7 +309,22 @@ func {{.Prefix}}ConnectionSetOption(cnxn *C.struct_AdbcConnection, key, val *C.c
 	}
 	conn := getFromHandle[cConn](cnxn.private_data)
 
-	rawCode := errToAdbcErr(err, conn.cnxn.(adbc.PostInitOptions).SetOption(C.GoString(key), C.GoString(val)))
+	if conn.cnxn == nil {
+		// not yet initialized
+		k, v := C.GoString(key), C.GoString(val)
+		if conn.initArgs == nil {
+			conn.initArgs = map[string]string{}
+		}
+		conn.initArgs[k] = v
+		return C.ADBC_STATUS_OK
+	}
+
+	opts, ok := conn.cnxn.(adbc.PostInitOptions)
+	if !ok {
+		setErr(err, "AdbcConnectionSetOption: not supported post-init")
+		return C.ADBC_STATUS_NOT_IMPLEMENTED
+	}
+	rawCode := errToAdbcErr(err, opts.SetOption(C.GoString(key), C.GoString(val)))
 	return C.AdbcStatusCode(rawCode)
 }
 
@@ -336,8 +352,25 @@ func {{.Prefix}}ConnectionInit(cnxn *C.struct_AdbcConnection, db *C.struct_AdbcD
 	if e != nil {
 		return C.AdbcStatusCode(errToAdbcErr(err, e))
 	}
-
 	conn.cnxn = c
+
+	if len(conn.initArgs) > 0 {
+		// C allow SetOption before Init, Go doesn't allow options to Open so set them now
+		opts, ok := conn.cnxn.(adbc.PostInitOptions)
+		if !ok {
+			setErr(err, "AdbcConnectionInit: options are not supported")
+			return C.ADBC_STATUS_NOT_IMPLEMENTED
+		}
+
+		for k, v := range conn.initArgs {
+			rawCode := errToAdbcErr(err, opts.SetOption(k, v))
+			if rawCode != adbc.StatusOK {
+				return C.AdbcStatusCode(rawCode)
+			}
+		}
+		conn.initArgs = nil
+	}
+
 	return C.ADBC_STATUS_OK
 }
 

--- a/go/adbc/pkg/flightsql/driver.go
+++ b/go/adbc/pkg/flightsql/driver.go
@@ -247,7 +247,8 @@ func FlightSQLDatabaseRelease(db *C.struct_AdbcDatabase, err *C.struct_AdbcError
 }
 
 type cConn struct {
-	cnxn adbc.Connection
+	cnxn     adbc.Connection
+	initArgs map[string]string
 }
 
 func checkConnAlloc(cnxn *C.struct_AdbcConnection, err *C.struct_AdbcError, fname string) bool {
@@ -312,7 +313,22 @@ func FlightSQLConnectionSetOption(cnxn *C.struct_AdbcConnection, key, val *C.cch
 	}
 	conn := getFromHandle[cConn](cnxn.private_data)
 
-	rawCode := errToAdbcErr(err, conn.cnxn.(adbc.PostInitOptions).SetOption(C.GoString(key), C.GoString(val)))
+	if conn.cnxn == nil {
+		// not yet initialized
+		k, v := C.GoString(key), C.GoString(val)
+		if conn.initArgs == nil {
+			conn.initArgs = map[string]string{}
+		}
+		conn.initArgs[k] = v
+		return C.ADBC_STATUS_OK
+	}
+
+	opts, ok := conn.cnxn.(adbc.PostInitOptions)
+	if !ok {
+		setErr(err, "AdbcConnectionSetOption: not supported post-init")
+		return C.ADBC_STATUS_NOT_IMPLEMENTED
+	}
+	rawCode := errToAdbcErr(err, opts.SetOption(C.GoString(key), C.GoString(val)))
 	return C.AdbcStatusCode(rawCode)
 }
 
@@ -340,8 +356,25 @@ func FlightSQLConnectionInit(cnxn *C.struct_AdbcConnection, db *C.struct_AdbcDat
 	if e != nil {
 		return C.AdbcStatusCode(errToAdbcErr(err, e))
 	}
-
 	conn.cnxn = c
+
+	if len(conn.initArgs) > 0 {
+		// C allow SetOption before Init, Go doesn't allow options to Open so set them now
+		opts, ok := conn.cnxn.(adbc.PostInitOptions)
+		if !ok {
+			setErr(err, "AdbcConnectionInit: options are not supported")
+			return C.ADBC_STATUS_NOT_IMPLEMENTED
+		}
+
+		for k, v := range conn.initArgs {
+			rawCode := errToAdbcErr(err, opts.SetOption(k, v))
+			if rawCode != adbc.StatusOK {
+				return C.AdbcStatusCode(rawCode)
+			}
+		}
+		conn.initArgs = nil
+	}
+
 	return C.ADBC_STATUS_OK
 }
 

--- a/go/adbc/pkg/panicdummy/driver.go
+++ b/go/adbc/pkg/panicdummy/driver.go
@@ -247,7 +247,8 @@ func PanicDummyDatabaseRelease(db *C.struct_AdbcDatabase, err *C.struct_AdbcErro
 }
 
 type cConn struct {
-	cnxn adbc.Connection
+	cnxn     adbc.Connection
+	initArgs map[string]string
 }
 
 func checkConnAlloc(cnxn *C.struct_AdbcConnection, err *C.struct_AdbcError, fname string) bool {
@@ -312,7 +313,22 @@ func PanicDummyConnectionSetOption(cnxn *C.struct_AdbcConnection, key, val *C.cc
 	}
 	conn := getFromHandle[cConn](cnxn.private_data)
 
-	rawCode := errToAdbcErr(err, conn.cnxn.(adbc.PostInitOptions).SetOption(C.GoString(key), C.GoString(val)))
+	if conn.cnxn == nil {
+		// not yet initialized
+		k, v := C.GoString(key), C.GoString(val)
+		if conn.initArgs == nil {
+			conn.initArgs = map[string]string{}
+		}
+		conn.initArgs[k] = v
+		return C.ADBC_STATUS_OK
+	}
+
+	opts, ok := conn.cnxn.(adbc.PostInitOptions)
+	if !ok {
+		setErr(err, "AdbcConnectionSetOption: not supported post-init")
+		return C.ADBC_STATUS_NOT_IMPLEMENTED
+	}
+	rawCode := errToAdbcErr(err, opts.SetOption(C.GoString(key), C.GoString(val)))
 	return C.AdbcStatusCode(rawCode)
 }
 
@@ -340,8 +356,25 @@ func PanicDummyConnectionInit(cnxn *C.struct_AdbcConnection, db *C.struct_AdbcDa
 	if e != nil {
 		return C.AdbcStatusCode(errToAdbcErr(err, e))
 	}
-
 	conn.cnxn = c
+
+	if len(conn.initArgs) > 0 {
+		// C allow SetOption before Init, Go doesn't allow options to Open so set them now
+		opts, ok := conn.cnxn.(adbc.PostInitOptions)
+		if !ok {
+			setErr(err, "AdbcConnectionInit: options are not supported")
+			return C.ADBC_STATUS_NOT_IMPLEMENTED
+		}
+
+		for k, v := range conn.initArgs {
+			rawCode := errToAdbcErr(err, opts.SetOption(k, v))
+			if rawCode != adbc.StatusOK {
+				return C.AdbcStatusCode(rawCode)
+			}
+		}
+		conn.initArgs = nil
+	}
+
 	return C.ADBC_STATUS_OK
 }
 

--- a/go/adbc/pkg/snowflake/driver.go
+++ b/go/adbc/pkg/snowflake/driver.go
@@ -247,7 +247,8 @@ func SnowflakeDatabaseRelease(db *C.struct_AdbcDatabase, err *C.struct_AdbcError
 }
 
 type cConn struct {
-	cnxn adbc.Connection
+	cnxn     adbc.Connection
+	initArgs map[string]string
 }
 
 func checkConnAlloc(cnxn *C.struct_AdbcConnection, err *C.struct_AdbcError, fname string) bool {
@@ -312,7 +313,22 @@ func SnowflakeConnectionSetOption(cnxn *C.struct_AdbcConnection, key, val *C.cch
 	}
 	conn := getFromHandle[cConn](cnxn.private_data)
 
-	rawCode := errToAdbcErr(err, conn.cnxn.(adbc.PostInitOptions).SetOption(C.GoString(key), C.GoString(val)))
+	if conn.cnxn == nil {
+		// not yet initialized
+		k, v := C.GoString(key), C.GoString(val)
+		if conn.initArgs == nil {
+			conn.initArgs = map[string]string{}
+		}
+		conn.initArgs[k] = v
+		return C.ADBC_STATUS_OK
+	}
+
+	opts, ok := conn.cnxn.(adbc.PostInitOptions)
+	if !ok {
+		setErr(err, "AdbcConnectionSetOption: not supported post-init")
+		return C.ADBC_STATUS_NOT_IMPLEMENTED
+	}
+	rawCode := errToAdbcErr(err, opts.SetOption(C.GoString(key), C.GoString(val)))
 	return C.AdbcStatusCode(rawCode)
 }
 
@@ -340,8 +356,25 @@ func SnowflakeConnectionInit(cnxn *C.struct_AdbcConnection, db *C.struct_AdbcDat
 	if e != nil {
 		return C.AdbcStatusCode(errToAdbcErr(err, e))
 	}
-
 	conn.cnxn = c
+
+	if len(conn.initArgs) > 0 {
+		// C allow SetOption before Init, Go doesn't allow options to Open so set them now
+		opts, ok := conn.cnxn.(adbc.PostInitOptions)
+		if !ok {
+			setErr(err, "AdbcConnectionInit: options are not supported")
+			return C.ADBC_STATUS_NOT_IMPLEMENTED
+		}
+
+		for k, v := range conn.initArgs {
+			rawCode := errToAdbcErr(err, opts.SetOption(k, v))
+			if rawCode != adbc.StatusOK {
+				return C.AdbcStatusCode(rawCode)
+			}
+		}
+		conn.initArgs = nil
+	}
+
 	return C.ADBC_STATUS_OK
 }
 

--- a/python/adbc_driver_flightsql/tests/conftest.py
+++ b/python/adbc_driver_flightsql/tests/conftest.py
@@ -24,23 +24,37 @@ import adbc_driver_flightsql.dbapi
 import adbc_driver_manager
 
 
-@pytest.fixture
-def dremio_uri():
+@pytest.fixture(scope="session")
+def dremio_uri() -> str:
     dremio_uri = os.environ.get("ADBC_DREMIO_FLIGHTSQL_URI")
     if not dremio_uri:
         pytest.skip("Set ADBC_DREMIO_FLIGHTSQL_URI to run tests")
-    yield dremio_uri
+    return dremio_uri
+
+
+@pytest.fixture(scope="session")
+def dremio_user() -> str:
+    username = os.environ.get("ADBC_DREMIO_FLIGHTSQL_USER")
+    if not username:
+        pytest.skip("Set ADBC_DREMIO_FLIGHTSQL_USER to run tests")
+    return username
+
+
+@pytest.fixture(scope="session")
+def dremio_pass() -> str:
+    password = os.environ.get("ADBC_DREMIO_FLIGHTSQL_PASS")
+    if not password:
+        pytest.skip("Set ADBC_DREMIO_FLIGHTSQL_PASS to run tests")
+    return password
 
 
 @pytest.fixture
-def dremio(dremio_uri):
-    username = os.environ.get("ADBC_DREMIO_FLIGHTSQL_USER")
-    password = os.environ.get("ADBC_DREMIO_FLIGHTSQL_PASS")
+def dremio(dremio_uri, dremio_user, dremio_pass):
     with adbc_driver_flightsql.connect(
         dremio_uri,
         db_kwargs={
-            adbc_driver_manager.DatabaseOptions.USERNAME.value: username,
-            adbc_driver_manager.DatabaseOptions.PASSWORD.value: password,
+            adbc_driver_manager.DatabaseOptions.USERNAME.value: dremio_user,
+            adbc_driver_manager.DatabaseOptions.PASSWORD.value: dremio_pass,
         },
     ) as db:
         with adbc_driver_manager.AdbcConnection(db) as conn:
@@ -48,14 +62,12 @@ def dremio(dremio_uri):
 
 
 @pytest.fixture
-def dremio_dbapi(dremio_uri):
-    username = os.environ.get("ADBC_DREMIO_FLIGHTSQL_USER")
-    password = os.environ.get("ADBC_DREMIO_FLIGHTSQL_PASS")
+def dremio_dbapi(dremio_uri, dremio_user, dremio_pass):
     with adbc_driver_flightsql.dbapi.connect(
         dremio_uri,
         db_kwargs={
-            adbc_driver_manager.DatabaseOptions.USERNAME.value: username,
-            adbc_driver_manager.DatabaseOptions.PASSWORD.value: password,
+            adbc_driver_manager.DatabaseOptions.USERNAME.value: dremio_user,
+            adbc_driver_manager.DatabaseOptions.PASSWORD.value: dremio_pass,
         },
     ) as conn:
         yield conn

--- a/python/adbc_driver_flightsql/tests/test_dbapi.py
+++ b/python/adbc_driver_flightsql/tests/test_dbapi.py
@@ -17,6 +17,9 @@
 
 import pyarrow
 
+import adbc_driver_flightsql.dbapi
+import adbc_driver_manager
+
 
 def test_query_trivial(dremio_dbapi):
     with dremio_dbapi.cursor() as cur:
@@ -32,3 +35,18 @@ def test_query_partitioned(dremio_dbapi):
 
         cur.adbc_read_partition(partitions[0])
         assert cur.fetchone() == (1,)
+
+
+def test_set_options(dremio_uri, dremio_user, dremio_pass):
+    # Regression test for apache/arrow-adbc#713
+    with adbc_driver_flightsql.dbapi.connect(
+        dremio_uri,
+        db_kwargs={
+            adbc_driver_manager.DatabaseOptions.USERNAME.value: dremio_user,
+            adbc_driver_manager.DatabaseOptions.PASSWORD.value: dremio_pass,
+        },
+        conn_kwargs={
+            "adbc.flight.sql.rpc.call_header.x-foo": "1",
+        },
+    ):
+        pass

